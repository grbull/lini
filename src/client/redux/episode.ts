/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
