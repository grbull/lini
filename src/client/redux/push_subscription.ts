/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { api } from '../utils/api';

type PushSubscriptionState = {
  status: 'loading' | 'idle' | 'error';
  error?: string;
};

const initialState: PushSubscriptionState = {
  status: 'idle',
};

export const create = createAsyncThunk<void, PushSubscriptionCreateDto>(
  'push_subscription/get',
  (createDto) => api.push.create(createDto)
);

export const pushSubscription = createSlice({
  name: 'push_subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state) => {
      state.status = 'idle';
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const pushSubscriptionActions = {
  ...pushSubscription.actions,
  create,
};
