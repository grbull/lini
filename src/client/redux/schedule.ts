/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type ScheduleState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: {
    future: EpisodeDto[];
    past: EpisodeDto[];
  };
  error?: string;
};

const initialState: ScheduleState = {
  status: 'init',
  data: {
    future: [],
    past: [],
  },
};

export const get = createAsyncThunk<EpisodeScheduleDto>('schedule/get', () =>
  api.episode.getSchedule()
);

export const schedule = createSlice({
  name: 'schedule',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const scheduleActions = {
  ...schedule.actions,
  get,
};
