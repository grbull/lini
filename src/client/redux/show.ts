/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowWithEpisodesDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type ShowState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data?: ShowWithEpisodesDto;
  error?: string;
};

const initialState: ShowState = {
  status: 'init',
};

export const get = createAsyncThunk<ShowWithEpisodesDto, number>(
  'show/get',
  (id) => api.show.get(id)
);

export const show = createSlice({
  name: 'show',
  initialState,
  reducers: {
    reset: () => {
      return initialState;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.data = undefined;
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const showActions = {
  ...show.actions,
  get,
};
