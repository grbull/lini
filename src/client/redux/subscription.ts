/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
} from '../../server/subscription/subscription.dto';
import { api } from '../utils/api';
import { scheduleActions } from './schedule';

type SubscriptionState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: SubscriptionDto[];
  error?: string;
};

const initialState: SubscriptionState = {
  status: 'init',
  data: [],
};

export const getAll = createAsyncThunk<SubscriptionDto[]>(
  'subscription/getAll',
  () => api.subscription.getAll()
);

export const create = createAsyncThunk<SubscriptionDto, SubscriptionCreateDto>(
  'subscription/create',
  async (createDto, thunkAPI) => {
    const subscription = await api.subscription.create(createDto);
    thunkAPI.dispatch(scheduleActions.get());
    return subscription;
  }
);

export const remove = createAsyncThunk<SubscriptionDto, SubscriptionRemoveDto>(
  'subscription/remove',
  async (removeDto, thunkAPI) => {
    const subscription = await api.subscription.remove(removeDto);
    thunkAPI.dispatch(scheduleActions.get());
    return subscription;
  }
);

export const subscription = createSlice({
  name: 'subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getAll.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(getAll.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(getAll.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });

    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.push(payload);
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });

    builder.addCase(remove.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(remove.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.splice(
        state.data.findIndex((s) => s.show.id === payload.show.id)
      );
    });
    builder.addCase(remove.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const subscriptionActions = {
  ...subscription.actions,
  getAll,
  create,
  remove,
};
