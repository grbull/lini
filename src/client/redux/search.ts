/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type SearchState = {
  isLoading: boolean;
  error?: string;
  data?: ShowDto[];
};

const initialState: SearchState = {
  isLoading: false,
};

export const query = createAsyncThunk<ShowDto[], string>(
  'search/query',
  (query) => api.show.search(query)
);

export const search = createSlice({
  name: 'search',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(query.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(query.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(query.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const searchActions = { ...search.actions, query };
