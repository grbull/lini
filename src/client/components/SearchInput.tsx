import cn from 'classnames';
import React, { ChangeEvent, ReactElement, useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';

import { searchActions } from '../redux/search';

export function SearchInput(): ReactElement {
  const [query, setQuery] = useState('');
  const dispatch = useDispatch();

  function changeHandler(e: ChangeEvent<HTMLInputElement>): void {
    setQuery(e.target.value);
  }

  useEffect(() => {
    if (query === '') {
      dispatch(searchActions.reset());
      return;
    }
    const timer = setTimeout(() => {
      dispatch(searchActions.query(query));
    }, 500);

    return (): void => clearTimeout(timer);
  }, [query, dispatch]);

  return (
    <div className={cn('px-2')}>
      <input
        aria-label="search-input"
        // eslint-disable-next-line jsx-a11y/no-autofocus
        autoFocus
        className={cn(
          'w-full',
          'mb-4',
          'p-1',
          'text-xl',
          'text-center',
          'bg-gray-200 dark:bg-gray-600',
          'border',
          'border-solid',
          'border-gray-800',
          'focus:border-gray-400'
        )}
        onChange={changeHandler}
        placeholder="Search..."
        type="search"
        value={query}
      />
    </div>
  );
}
