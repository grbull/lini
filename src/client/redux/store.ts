import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { episode } from './episode';
import { pushSubscription } from './push_subscription';
import { schedule } from './schedule';
import { search } from './search';
import { show } from './show';
import { subscription } from './subscription';
import { user } from './user';

export const reducers = combineReducers({
  user: user.reducer,
  search: search.reducer,
  subscription: subscription.reducer,
  show: show.reducer,
  episode: episode.reducer,
  schedule: schedule.reducer,
  pushSubscription: pushSubscription.reducer,
});

const persistConfig = {
  key: 'root',
  storage,
  version: 2,
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  // https://github.com/rt2zz/redux-persist/issues/988#issuecomment-552242978
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
