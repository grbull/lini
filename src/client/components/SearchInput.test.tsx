/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { configureStore } from '@reduxjs/toolkit';
import {
  fireEvent,
  render,
  RenderResult,
  waitFor,
} from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import { reducers } from '../redux/store';
import { SearchInput } from './SearchInput';

function setup(): RenderResult & { input: HTMLInputElement } {
  const store = configureStore({
    reducer: reducers,
  });

  const utils = render(
    <Provider store={store}>
      <SearchInput />
    </Provider>
  );

  const input = utils.getByLabelText('search-input') as HTMLInputElement;

  return { input, ...utils };
}

describe('SearchInput Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles input change', async () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: 'query' } });
    fireEvent.focusOut(input);

    await waitFor(() => expect(input.value).toBe('query'));
  });
});
