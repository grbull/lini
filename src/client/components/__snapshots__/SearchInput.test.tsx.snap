// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-2"
  >
    <input
      aria-label="search-input"
      class="w-full mb-4 p-1 text-xl text-center bg-gray-200 dark:bg-gray-600 border border-solid border-gray-800 focus:border-gray-400"
      placeholder="Search..."
      type="search"
      value=""
    />
  </div>
</DocumentFragment>
`;
